man/rzxcheck.1
.\" -*- nroff -*-
.\"
.\" rzxcheck.1: rzxdump man page
.\" Copyright (c) 2003-2004 Philip Kendall
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" Author contact information:
.\"
.\" E-mail: philip-fuse@shadowmagic.org.uk
.\"
.\"
.TH rzxcheck 1 "24th May, 2013" "Version 1.1.1" "Emulators"
.\"
.\"------------------------------------------------------------------
.\"
.SH NAME
rzxcheck \(em Check the signature in a ZX Spectrum RZX file
.\"
.\"------------------------------------------------------------------
.\"
.SH SYNOPSIS
.B rzxcheck
.IR "rzxfile" ...
.P
.\"
.\"------------------------------------------------------------------
.\"
.SH DESCRIPTION
rzxcheck verifies the digital signature found in a Sinclair ZX
Spectrum RZX input recording file.
.\"
.\"------------------------------------------------------------------
.\"
.SH OPTIONS
.TP
.I rzxfile
specifies an input recording file whose signature is to be verified.
.\"
.\"------------------------------------------------------------------
.\"
.SH BUGS
None known.
.\"
.\"------------------------------------------------------------------
.\"
.SH SEE ALSO
.IR fuse "(1),"
.IR fuse\-utils "(1),"
.IR rzxdump "(1),"
.IR rzxtool "(1)"
.PP
The comp.sys.sinclair Spectrum FAQ, at
.br
.IR "http://www.worldofspectrum.org/faq/index.html" .
.\"
.\"------------------------------------------------------------------
.\"
.SH AUTHOR
Philip Kendall (philip\-fuse@shadowmagic.org.uk).


ChangeLog
2013-05-xx  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.1 released.

        * Decreased libtool version to allow binaries built against
          libspectrum 1.0.0 to run with libspectrum 1.1.1.

2013-05-18  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.1.0 released.

        * Support for Spectranet in snapshots (Philip Kendall).

        * Support for Cheetah SpecDrum in snapshots (Fredrick Meunier).

        * Support reading PZX files (Fredrick Meunier).

        * Support for the TZX set signal level block (Fredrick Meunier).

        * Various minor bug fixes/improvements:
          * Add support for detecting POK files (Sergio Baldoví).
          * Advertise support for bzip2 compressed and wav files (Sergio
            Baldoví).
          * Drop support for GLib version 1.x (Alberto Garcia).
          * Add support for audiofile 0.3.x (Alberto Garcia).
          * Load all snapshot RAM pages for machines with more than 128k 
            (Fredrick Meunier; thanks, ketmar).
          * Support more SZX 1.5 features (Fredrick Meunier).
          * Modernise autoconf support (Sergio Baldoví).

2010-12-16  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 1.0.0 released.

        * New end of tape flag (Fredrick Meunier).

        * AF and AF' were swapped when reading and writing SZX files
          (Fredrick Meunier; thanks, velesoft).

        * Support for Fuller Box and Melodik (and similar AY interfaces) in
          snapshots (Fredrick Meunier).

        * Support for Opus Discovery in snapshots (Fredrick Meunier, Gergely
          Szasz).

        * Support for the NTSC 48K Spectrum (Philip Kendall).

        * Support for stereo sound files (Fredrick Meunier; thanks, Gilberto
          Almeida).

        * Add proper support for +2A/+3, Pentagon and Scorpion in .z80 files
          (Philip Kendall; thanks, Nick Corey)

        * Various fixes for compiling with Visual Studio 6 (Simon Owen).

        * Various minor bug fixes/improvements:
          * Looser signature checks for .dsk files (Gergely Szasz; thanks,
            zx81 and Simon Owen).
          * --without-(bzip2|zlib) options to configure for Gentoo and similar
            distributions (José Manuel Ferrer Ortiz).
          * Handle querying the block length for turbo and pure data blocks
            with 0 length (Fredrick Meunier).
          * Fix loading of Warajevo TAPs with raw samples (Fredrick Meunier).
          * Fail sensibly if file decompression fails (Fredrick Meunier).
          * Remove various unused variables, initialisations etc (Fredrick
            Meunier).

2008-11-30  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

        * libspectrum 0.5.0 released.

	* Memory management change: memory allocation functions now abort
	  on failure. Leads to simpler code both internally and for
	  users of the library (Philip Kendall).

	* Support for tape acceleration (Philip Kendall)

	* IDE improvements to allow DivIDE to support FATware (Philip
	  Kendall; thanks, http://www.dusky.sk/zxs/)

	* New API to deal with flagging automatic snapshots in RZX files
	  (Philip Kendall).

	* New function to get length (time) of a tape block (Philip Kendall,
	  Fredrick Meunier).

	* 128K SNA files are now treated as being for the Pentagon 128
	  rather than the Spectrum 128 (Fredrick Meunier)

	* SNA files now written out correctly (Philip Kendall).

	* Handle MDR write protection flag correctly (Philip Kendall)

	* Allow MDR images with bad checksums to load (the error will
	  be handled by the If1 ROM) (Gergely Szasz)

        * Support saving custom ROMs in SZX snapshots (Fredrick Meunier,
	  Stuart Brady)

	* Support for DivIDE, Kempston mouse and the simple IDE interface
	  in SZX snapshots (Fredrick Meunier)

        * Allow building without libgcrypt (Stuart Brady)

	* Amiga and MorphOS compilation fixes (Chris Young, Q-Master)

	* Don't produce empty libraries to fix build process on
	  OpenSolaris (Fredrick Meunier)

	* Remove deprecated functions (Fredrick Meunier).

	* Tweak timings for TS2068 (Fredrick Meunier; thanks Philip Kendall).

        * NTSC Spectrum support (Philip Kendall; thanks, Claudio Bernet).

2008-01-05  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.4.0 released.

	* Add support for loading from WAV files (Fredrick Meunier).

	* Add support for CSW tape files (Fredrick Meunier).

	* Tweak timings for TC2048 and TC2068 (Fredrick Meunier; thanks Philip
	  Kendall and Mark Woodmass).

	* Add support for the +D in z80 and szx snapshots (Fredrick Meunier).

	* Add Pentagon 512/1024 support (Q-Master).

	* Allow Amiga version to decompress files with xfdmaster.library
	  (Chris Young).

	* Allow MorphOS version to use xfdmaster.library (Q-Master).

	* Various other bugfixes.

2007-05-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0.1 released (bug fix release).

	* Avoid crash when reading an invalid TZX file (thanks, crabfists).
	  (Philip Kendall).

	* Export the TZX Generalized Data Block functions when building a
	  Win32 DLL (Stuart Brady).

	* Distribute the hacking/ directory (thanks, Stuart Brady) (Philip
	  Kendall).

	* Remove warning when building from CVS (Stuart Brady).

2007-04-11  Philip Kendall  <philip-fuse@shadowmagic.org.uk>

	* libspectrum 0.3.0 released.

	* Support for RZX rollback, including having snapshots and
	  multiple input recording blocks in input recording files (Philip
	  Kendall).

	* Improved IDE emulation (Matthew Westcott).

	* Support for microdrive images (Philip Kendall, Gergely Szasz).

	* TZX Generalized Data Block support. (Philip Kendall).

	* Support for Z80Em and CSW tape files (Darren Salt).

	* Support for the TS2068 (Fredrick Meunier).

	* Support for connected joystick information in snapshots (Fredrick
	  Meunier).

	* Add interrupt length to timing information (Philip Kendall).


rzxcheck.c
/* rzxcheck.c: Check the signature on an RZX file
   Copyright (c) 2002-2003 Philip Kendall

   $Id$

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with this program; if not, write to the Free Software Foundation, Inc.,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Author contact information:

   E-mail: philip-fuse@shadowmagic.org.uk

*/

#include <config.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <libspectrum.h>

#include "utils.h"

char *progname;			/* argv[0] */

int
main( int argc, char **argv )
{
  unsigned char *buffer; size_t length;

  const char *rzxfile;

  libspectrum_rzx *rzx;

  libspectrum_error error;
  libspectrum_dword keyid = 0;
  libspectrum_signature signature;
  struct rzx_key *key;

  progname = argv[0];

  if( init_libspectrum() ) return 16;

  if( argc < 2 ) {
    fprintf( stderr, "%s: usage: %s <rzxfile>\n", progname, progname );
    return 2;
  }
  rzxfile = argv[1];

  rzx = libspectrum_rzx_alloc();

  if( read_file( rzxfile, &buffer, &length ) ) return 16;

  if( libspectrum_rzx_read( rzx, buffer, length ) ) {
    free( buffer );
    return 16;
  }

  keyid = libspectrum_rzx_get_keyid( rzx );
  if( !keyid ) {
    printf( "%s: no key ID found in '%s'\n", progname, rzxfile );
    libspectrum_rzx_free( rzx );
    free( buffer );
    return 16;
  }

  for( key = known_keys; key->id; key++ )
    if( keyid == key->id ) break;

  if( !key->id ) {
    printf( "%s: don't know anything about key ID %08x\n", progname,
	    keyid );
    libspectrum_rzx_free( rzx );
    free( buffer );
    return 16;
  }

  error = libspectrum_rzx_get_signature( rzx, &signature );
  if( error ) {
    libspectrum_rzx_free( rzx );
    free( buffer );
    return 16;
  }

  error = libspectrum_verify_signature( &signature, &( key->key ) );
  if( error && error != LIBSPECTRUM_ERROR_SIGNATURE ) {
    libspectrum_signature_free( &signature );
    libspectrum_rzx_free( rzx );
    free( buffer );
    return 16;
  }

  free( buffer );

  libspectrum_rzx_free( rzx ); free( rzx );
  libspectrum_signature_free( &signature );

  if( error == LIBSPECTRUM_ERROR_SIGNATURE ) {
    printf( "%s: BAD signature with key %08x (%s) in '%s'\n", progname,
	    key->id, key->description, rzxfile );
    return 1;
  } else {
    printf( "%s: good signature with key %08x (%s) in '%s'\n", progname,
	    key->id, key->description, rzxfile );
    return 0;
  }

}
